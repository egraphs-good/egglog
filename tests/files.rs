use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
