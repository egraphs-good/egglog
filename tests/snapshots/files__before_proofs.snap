---
source: tests/files.rs
expression: serialized
---
nodes:
  Rational-0:
    op: (rational 2 1)
    children: []
    eclass: Rational-0
    cost: 0
  RepConst_Rational__-0:
    op: RepConst_Rational__
    children:
      - Rational-0
    eclass: "20"
    cost: 1
  Rational-5871781006564002453:
    op: (rational 0 1)
    children: []
    eclass: Rational-5871781006564002453
    cost: 0
  RepConst_Rational__-5871781006564002453:
    op: RepConst_Rational__
    children:
      - Rational-5871781006564002453
    eclass: "60"
    cost: 1
  Original__-5871781006564002453:
    op: Original__
    children:
      - TrmOf__-17615343019692007359
    eclass: "2"
    cost: 1
  Original__-4208978898528913939:
    op: Original__
    children:
      - TrmOf__-15952540911656918845
    eclass: "8"
    cost: 1
  Original__-3377577844511369682:
    op: Original__
    children:
      - TrmOf__-15121139857639374588
    eclass: "11"
    cost: 1
  Original__-16888066223598386617:
    op: Original__
    children:
      - TrmOf__-987321939341122219
    eclass: "403"
    cost: 1
  Original__-5040733954629534610:
    op: Original__
    children: []
    eclass: "417"
    cost: 1
  Original__-1403464968953739865:
    op: Original__
    children: []
    eclass: "5"
    cost: 1
  Original__-16784295967757539516:
    op: Original__
    children:
      - AstVar_String__-3377577844511369682
    eclass: "15"
    cost: 1
  Original__-17511749764892698465:
    op: Original__
    children: []
    eclass: "19"
    cost: 1
  Original__-9041995340435216140:
    op: Original__
    children: []
    eclass: "23"
    cost: 1
  Original__-13978428036099289028:
    op: Original__
    children: []
    eclass: "336"
    cost: 1
  RuleTerm__-1480107397956408125:
    op: RuleTerm__
    children:
      - TrmOf__-16004514540098111499
    eclass: "66"
    cost: 1
  RuleTerm__-68000147645774473:
    op: RuleTerm__
    children:
      - TrmOf__-11016285217034384164
    eclass: "88"
    cost: 1
  RuleTerm__-16113564009593019811:
    op: RuleTerm__
    children:
      - TrmOf__-16160435425421689461
    eclass: "223"
    cost: 1
  RuleTerm__-14909174272005618568:
    op: RuleTerm__
    children:
      - TrmOf__-1195216453105892835
    eclass: "150"
    cost: 1
  RuleTerm__-8632718619206876900:
    op: RuleTerm__
    children:
      - TrmOf__-9717121507046106021
    eclass: "208"
    cost: 1
  RuleTerm__-7050700098532748677:
    op: RuleTerm__
    children:
      - TrmOf__-5716037122281962698
    eclass: "289"
    cost: 1
  RuleTerm__-827329982148558853:
    op: RuleTerm__
    children:
      - TrmOf__-16004514540098111499
    eclass: "102"
    cost: 1
  RuleTerm__-4891502376648696847:
    op: RuleTerm__
    children:
      - TrmOf__-987321939341122219
    eclass: "167"
    cost: 1
  RuleTerm__-4296628794274152636:
    op: RuleTerm__
    children:
      - TrmOf__-16160435425421689461
    eclass: "275"
    cost: 1
  RuleTerm__-1464556065508684311:
    op: RuleTerm__
    children:
      - TrmOf__-11016285217034384164
    eclass: "306"
    cost: 1
  ProofNull__-0:
    op: ProofNull__
    children: []
    eclass: "0"
    cost: 1
  Null__-0:
    op: Null__
    children: []
    eclass: "0"
    cost: 9223372036854776000
  TrmOf__-15121139857639374588:
    op: TrmOf__
    children:
      - MakeTrmPrf__-2651793105796594534
    eclass: "10"
    cost: 1000000
  TrmOf__-15952540911656918845:
    op: TrmOf__
    children:
      - MakeTrmPrf__-15787446741851832475
    eclass: "7"
    cost: 1000000
  TrmOf__-17615343019692007359:
    op: TrmOf__
    children:
      - MakeTrmPrf__-7659469028595837896
    eclass: "1"
    cost: 1000000
  TrmOf__-16004514540098111499:
    op: TrmOf__
    children:
      - MakeTrmPrf__-3804304252294174724
    eclass: "65"
    cost: 1000000
  TrmOf__-1195216453105892835:
    op: TrmOf__
    children:
      - MakeTrmPrf__-1262483645723449506
    eclass: "152"
    cost: 1000000
  TrmOf__-9717121507046106021:
    op: TrmOf__
    children:
      - MakeTrmPrf__-5699467722751906279
    eclass: "210"
    cost: 1000000
  TrmOf__-16160435425421689461:
    op: TrmOf__
    children:
      - MakeTrmPrf__-17122760709440669650
    eclass: "222"
    cost: 1000000
  TrmOf__-5716037122281962698:
    op: TrmOf__
    children:
      - MakeTrmPrf__-15035091326745903097
    eclass: "291"
    cost: 1000000
  TrmOf__-11016285217034384164:
    op: TrmOf__
    children:
      - MakeTrmPrf__-10201126904246574918
    eclass: "405"
    cost: 1000000
  TrmOf__-987321939341122219:
    op: TrmOf__
    children:
      - MakeTrmPrf__-14825894830305030089
    eclass: "405"
    cost: 1000000
  AstVar_String__-3377577844511369682:
    op: AstVar_String__
    children:
      - TrmOf__-15121139857639374588
    eclass: "716"
    cost: 1
  MakeTrmPrf__-7659469028595837896:
    op: MakeTrmPrf__
    children:
      - TrmOf__-17615343019692007359
      - Original__-5871781006564002453
    eclass: "3"
    cost: 1
  MakeTrmPrf__-15787446741851832475:
    op: MakeTrmPrf__
    children:
      - TrmOf__-15952540911656918845
      - Original__-4208978898528913939
    eclass: "9"
    cost: 1
  MakeTrmPrf__-2651793105796594534:
    op: MakeTrmPrf__
    children:
      - TrmOf__-15121139857639374588
      - Original__-3377577844511369682
    eclass: "12"
    cost: 1
  MakeTrmPrf__-3804304252294174724:
    op: MakeTrmPrf__
    children:
      - TrmOf__-16004514540098111499
      - RuleTerm__-1480107397956408125
    eclass: "31"
    cost: 1
  MakeTrmPrf__-14825894830305030089:
    op: MakeTrmPrf__
    children:
      - TrmOf__-987321939341122219
      - RuleTerm__-68000147645774473
    eclass: "63"
    cost: 1
  MakeTrmPrf__-17122760709440669650:
    op: MakeTrmPrf__
    children:
      - TrmOf__-16160435425421689461
      - RuleTerm__-16113564009593019811
    eclass: "97"
    cost: 1
  MakeTrmPrf__-1262483645723449506:
    op: MakeTrmPrf__
    children:
      - TrmOf__-1195216453105892835
      - RuleTerm__-14909174272005618568
    eclass: "151"
    cost: 1
  MakeTrmPrf__-5699467722751906279:
    op: MakeTrmPrf__
    children:
      - TrmOf__-9717121507046106021
      - RuleTerm__-8632718619206876900
    eclass: "209"
    cost: 1
  MakeTrmPrf__-15035091326745903097:
    op: MakeTrmPrf__
    children:
      - TrmOf__-5716037122281962698
      - RuleTerm__-7050700098532748677
    eclass: "290"
    cost: 1
  MakeTrmPrf__-10201126904246574918:
    op: MakeTrmPrf__
    children:
      - TrmOf__-11016285217034384164
      - Original__-16888066223598386617
    eclass: "404"
    cost: 1
  DemandEq__-5815586648995041194:
    op: DemandEq__
    children: []
    eclass: "444"
    cost: 1000000
  DemandEq__-7068423019940469303:
    op: DemandEq__
    children: []
    eclass: "989"
    cost: 1000000
  DemandEq__-2671660272381556966:
    op: DemandEq__
    children:
      - TrmOf__-16004514540098111499
    eclass: "155"
    cost: 1000000
  DemandEq__-6672567656104162082:
    op: DemandEq__
    children:
      - TrmOf__-987321939341122219
    eclass: "213"
    cost: 1000000
  DemandEq__-6568797400263314981:
    op: DemandEq__
    children:
      - TrmOf__-16160435425421689461
    eclass: "227"
    cost: 1000000
  DemandEq__-8231776509339941702:
    op: DemandEq__
    children: []
    eclass: "567"
    cost: 1000000
  DemandEq__-15870306880821417977:
    op: DemandEq__
    children: []
    eclass: "860"
    cost: 1000000
  DemandEq__-17948721015344509516:
    op: DemandEq__
    children: []
    eclass: "917"
    cost: 1000000
  DemandEq__-1544537807812465189:
    op: DemandEq__
    children:
      - AstVar_String__-3377577844511369682
      - AstVar_String__-3377577844511369682
    eclass: "444"
    cost: 1000000
  DemandEq__-2648223368006877125:
    op: DemandEq__
    children: []
    eclass: "444"
    cost: 1000000
  Congruence__-13691527104406518518:
    op: Congruence__
    children:
      - RuleTerm__-68000147645774473
    eclass: "159"
    cost: 1
  Congruence__-3766688084637180088:
    op: Congruence__
    children:
      - RuleTerm__-68000147645774473
    eclass: "217"
    cost: 1
  Congruence__-233012353760694237:
    op: Congruence__
    children:
      - RuleTerm__-68000147645774473
    eclass: "298"
    cost: 1
  Congruence__-17276645460599582402:
    op: Congruence__
    children:
      - RuleTerm__-68000147645774473
    eclass: "571"
    cost: 1
  Congruence__-4181946109042106699:
    op: Congruence__
    children:
      - RuleTerm__-68000147645774473
    eclass: "864"
    cost: 1
  Congruence__-12692547903445119682:
    op: Congruence__
    children: []
    eclass: "880"
    cost: 1
  Congruence__-6832070156418317432:
    op: Congruence__
    children:
      - RuleTerm__-68000147645774473
    eclass: "921"
    cost: 1
  Congruence__-17198619085230645144:
    op: Congruence__
    children: []
    eclass: "993"
    cost: 1
  Congruence__-7618609781449447259:
    op: Congruence__
    children:
      - RuleTerm__-1480107397956408125
    eclass: "111"
    cost: 1
  Congruence__-5746908166228460519:
    op: Congruence__
    children: []
    eclass: "132"
    cost: 1
root_eclasses: []
class_data:
  "20":
    type: TrmPrf__
  Rational-0:
    type: Rational
  "60":
    type: TrmPrf__
  Rational-5871781006564002453:
    type: Rational
  "2":
    type: Proof__
  "1":
    type: Ast__
  "8":
    type: Proof__
  "7":
    type: Ast__
  "11":
    type: Proof__
  "10":
    type: Ast__
  "403":
    type: Proof__
  "405":
    type: Ast__
  "417":
    type: Proof__
  "714":
    type: Ast__
  "5":
    type: Proof__
  "949":
    type: Ast__
  "15":
    type: Proof__
  "716":
    type: Ast__
  "19":
    type: Proof__
  "669":
    type: Ast__
  "23":
    type: Proof__
  "988":
    type: Ast__
  "336":
    type: Proof__
  "948":
    type: Ast__
  "66":
    type: Proof__
  "26":
    type: Proof__
  "65":
    type: Ast__
  "88":
    type: Proof__
  "33":
    type: Proof__
  "223":
    type: Proof__
  "92":
    type: Proof__
  "222":
    type: Ast__
  "150":
    type: Proof__
  "99":
    type: Proof__
  "152":
    type: Ast__
  "208":
    type: Proof__
  "164":
    type: Proof__
  "210":
    type: Ast__
  "289":
    type: Proof__
  "272":
    type: Proof__
  "291":
    type: Ast__
  "102":
    type: Proof__
  "167":
    type: Proof__
  "275":
    type: Proof__
  "306":
    type: Proof__
  "303":
    type: Proof__
  "0":
    type: ProofList__
  "12":
    type: TrmPrf__
  "9":
    type: TrmPrf__
  "3":
    type: TrmPrf__
  "31":
    type: TrmPrf__
  "151":
    type: TrmPrf__
  "209":
    type: TrmPrf__
  "97":
    type: TrmPrf__
  "290":
    type: TrmPrf__
  "404":
    type: TrmPrf__
  "63":
    type: TrmPrf__
  "444":
    type: Proof__
  "989":
    type: Proof__
  "930":
    type: Ast__
  "987":
    type: Ast__
  "155":
    type: Proof__
  "213":
    type: Proof__
  "227":
    type: Proof__
  "567":
    type: Proof__
  "801":
    type: Ast__
  "860":
    type: Proof__
  "858":
    type: Ast__
  "917":
    type: Proof__
  "915":
    type: Ast__
  "667":
    type: Ast__
  "159":
    type: Proof__
  "158":
    type: ProofList__
  "217":
    type: Proof__
  "216":
    type: ProofList__
  "298":
    type: Proof__
  "297":
    type: ProofList__
  "571":
    type: Proof__
  "570":
    type: ProofList__
  "864":
    type: Proof__
  "863":
    type: ProofList__
  "880":
    type: Proof__
  "266":
    type: Proof__
  "879":
    type: ProofList__
  "921":
    type: Proof__
  "920":
    type: ProofList__
  "993":
    type: Proof__
  "238":
    type: Proof__
  "992":
    type: ProofList__
  "111":
    type: Proof__
  "132":
    type: Proof__
  "123":
    type: Proof__

